contracts-bedrock/contracts/L1/ResourceMetering.sol

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.10;
+
+import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
+import { SignedMath } from "@openzeppelin/contracts/utils/math/SignedMath.sol";
+import { SlotPacking128x64x64 } from "../libraries/SlotPacking128x64x64.sol";
+import { Burn } from "../libraries/Burn.sol";
+
+/**
+ * @title ResourceMetering
+ * @notice ResourceMetering implements an EIP-1559 style resource metering system where pricing
+ * updates automatically based on current demand.
+ */
+contract ResourceMetering {
+    /**
+     * Along with the resource limit, determines the target resource limit.
+     */
+    int256 public constant ELASTICITY_MULTIPLIER = 2;
+
+    /**
+     * Denominator that determines max change on fee per block.
+     */
+    int256 public constant BASE_FEE_MAX_CHANGE_DENOMINATOR = 8;
+
+    /**
+     * Maximum amount of deposit gas that can be used within this block.
+     */
+    int256 public constant MAX_RESOURCE_LIMIT = 30000000;
+
+    /**
+     * Target amount of deposit gas that should be used within this block.
+     */
+    int256 public constant TARGET_RESOURCE_LIMIT = MAX_RESOURCE_LIMIT / ELASTICITY_MULTIPLIER;
+
+    /**
+     * Minimum base fee value, cannot go lower than this.
+     */
+    int256 public constant MINIMUM_BASE_FEE = 10000;
+
+    /**
+     * Initial base fee value.
+     */
+    uint128 public constant INITIAL_BASE_FEE = 1000000000;
```

Comment:
We will want to make sure that this initial base fee is not too high or the first few deposits will be prohibitively expensive